queue/back_pressure_test.go

```diff
@@ -0,0 +1,375 @@
+		minThresh, maxThresh, WithRandSource(localRng.Float64),
+	)
+
+	// Next, we'll define a custom predicate: drop items with value 42.
+	customValuePredicate := func(queueLen int, item int) bool {
+		return item == 42
+	}
+
+	// We'll also make a composed predicate: drop if RED says so OR if item
+	// is 42.
+	composedPredicate := func(queueLen int, item int) bool {
+		isRedDrop := redPred(queueLen, item)
+		isCustomDrop := customValuePredicate(queueLen, item)
+		return isRedDrop || isCustomDrop
+	}
+
+	q := NewBackpressureQueue[int](capacity, composedPredicate)
+
+	// Scenario 1: Enqueue item 42 when queue length is between min/max
+	// thresholds. As we're below the max threshold, we shouldn't drop
+	// anything.
+	for i := 0; i < minThresh; i++ {
+		// All items aren't 42, so they shouldn't be dropped.
+		res := q.Enqueue(context.Background(), i)
+		require.True(
+			t, res.IsOk(), "enqueue S1 setup item %d Ok: %v", i,
+			res.Err(),
+		)
+
+		droppedVal, _ := res.Unpack()
+		require.False(
+			t, droppedVal, "enqueue S1 setup item %d (qLen "+
+				"before: %d) should not be dropped. Predicate "+
+				"was redPred(%d,%d) || customPred(%d,%d)", i,
+			len(q.ch)-1, len(q.ch)-1, i, len(q.ch)-1, i,
+		)
+	}
+
+	currentLen := len(q.ch)
+	require.Equal(t, minThresh, currentLen, "queue length after S1 setup")
+
+	// Enqueue item 42. customValuePredicate is true, so composedPredicate
+	// is true. Item 42 should be dropped regardless of what redPred
+	// decides.
+	res := q.Enqueue(context.Background(), 42)
+	require.True(
+		t, res.IsOk(), "Enqueue of 42 should not error: %v", res.Err(),
+	)
+	require.True(t,
+		res.UnwrapOrFail(t), "Item 42 should have been dropped by "+
+			"composed predicate",
+	)
+	require.Equal(
+		t, currentLen, len(q.ch), "queue length should not change "+
+			"after dropping 42",
+	)
+
+	// Scenario 2: Queue is full. Item 100 (not 42). Reset the localRng
+	// state by re-seeding for the setup of Scenario 2 to ensure its
+	// behavior is independent of S1.
+	localRng.Seed(testSeed + 1)
+
+	// The goal is to get it to capacity for the next step. First, create a
+	// temporary queue with only redPred to observe its behavior or fill.
+	// This part is mostly for potential debugging or complex fill logic.
+	// For this test, we'll simplify by directly filling the SUT queue.
+	localRng.Seed(testSeed + 2)
+
+	// Re-create the main SUT queue with the composedPredicate. We will
+	// manually fill its channel to capacity to bypass Enqueue logic for
+	// setup.
+	q = NewBackpressureQueue[int](capacity, composedPredicate)
+	for i := 0; i < capacity; i++ {
+		// Directly add items to the channel, bypassing Enqueue logic
+		// for this setup. Ensure items are not 42, so
+		// customValuePredicate is false for them. Behavior of redPred
+		// part is controlled by localRng.
+		q.ch <- i
+	}
+	require.Equal(
+		t, capacity, len(q.ch), "queue manually filled to capacity "+
+			"for S2 test",
+	)
+
+	localRng.Seed(testSeed + 3)
```

Comment:
We really only need to reseed once (and even that's debatable). But def not three times. The rnd only comes into play when we hit the predicate, which happens in the line below.